ADAL/src/urlprotocol/ADURLSessionDemux.m

```diff
@@ -0,0 +1,234 @@
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "ADURLSessionDemux.h"
+
+
+@interface ADURLSessionDemuxTaskInfo : NSObject
+
+- (instancetype)initWithTask:(NSURLSessionDataTask *)task delegate:(id<NSURLSessionDataDelegate>)delegate;
+
+@property (atomic, strong) NSURLSessionDataTask *task;
+@property (atomic, strong) id<NSURLSessionDataDelegate> delegate;
+@property (atomic, strong) NSThread *thread;
+
+- (void)performBlock:(dispatch_block_t)block;
+
+- (void)invalidate;
+
+@end
+
+@implementation ADURLSessionDemuxTaskInfo
+
+- (instancetype)initWithTask:(NSURLSessionDataTask *)task delegate:(id<NSURLSessionDataDelegate>)delegate
+{
+    self = [super init];
+    if (self != nil)
+    {
+        self->_task = task;
+        self->_delegate = delegate;
+        self->_thread = [NSThread currentThread];
+    }
+    return self;
+}
+
+- (void)performBlock:(dispatch_block_t)block
+{
+    [self performSelector:@selector(performBlockOnClientThread:)
+                 onThread:self.thread
+               withObject:[block copy]
+            waitUntilDone:NO];
+}
+
+- (void)performBlockOnClientThread:(dispatch_block_t)block
+{
+    block();
+}
+
+- (void)invalidate
+{
+    self.delegate = nil;
+    self.thread = nil;
+}
+
+@end
+
+
+@interface ADURLSessionDemux() <NSURLSessionDataDelegate>
+
+@property (atomic, strong, readonly) NSMutableDictionary *taskInfoByTaskID;
+@property (atomic, strong, readonly) NSOperationQueue *sessionDelegateQueue;
+
+@end
+
+
+@implementation ADURLSessionDemux
+
+- (instancetype)initWithConfiguration:(NSURLSessionConfiguration *)configuration
+{
+    self = [super init];
+    if (self)
+    {
+        self->_configuration = [configuration copy];
+        
+        self->_taskInfoByTaskID = [[NSMutableDictionary alloc] init];
+        
+        self->_sessionDelegateQueue = [[NSOperationQueue alloc] init];
+        [self->_sessionDelegateQueue setMaxConcurrentOperationCount:1];
+        [self->_sessionDelegateQueue setName:@"ADURLSessionDemux"];
```

Comment:
It is being used at exactly 2 places, and they are separated by 2 lines.
I am open to making this constant but not convinced it adds any benefit here.